pkg/database/database_integration_test.go

```diff
@@ -0,0 +1,171 @@
+package database_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"mime/multipart"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	awss3 "github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/dhis2-sre/im-manager/pkg/database"
+	"github.com/dhis2-sre/im-manager/pkg/inttest"
+	"github.com/dhis2-sre/im-manager/pkg/model"
+	"github.com/dhis2-sre/im-manager/pkg/storage"
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDatabaseHandler(t *testing.T) {
+	t.Parallel()
+
+	db := inttest.SetupDB(t)
+
+	s3Dir := t.TempDir()
+	s3Bucket := "database-bucket"
+	err := os.Mkdir(s3Dir+"/"+s3Bucket, 0o755)
+	require.NoError(t, err, "failed to create S3 output bucket")
+	s3 := inttest.SetupS3(t, s3Dir)
+	uploader := manager.NewUploader(s3.Client)
+	s3Client := storage.NewS3Client(s3.Client, uploader)
+
+	databaseRepository := database.NewRepository(db)
+	databaseService := database.NewService(s3Bucket, s3Client, groupService{}, databaseRepository)
+
+	client := inttest.SetupHTTPServer(t, func(engine *gin.Engine) {
+		databaseHandler := database.NewHandler(databaseService, groupService{groupName: "packages"}, instanceService{}, stackService{})
+		database.Routes(engine, func(ctx *gin.Context) {
+			ctx.Set("user", &model.User{
+				Email: "user1@dhis2.org",
+				Groups: []model.Group{
+					{
+						Name: "packages",
+					},
+				},
+			})
+		}, databaseHandler)
+	})
+
+	var databaseID string
+	{
+		t.Log("Upload")
+
+		var b bytes.Buffer
+		w := multipart.NewWriter(&b)
+		err := w.WriteField("group", "packages")
+		require.NoError(t, err, "failed to write form field")
+		f, err := w.CreateFormFile("database", "mydb")
+		require.NoError(t, err, "failed to create form file")
+		_, err = io.WriteString(f, "file contents")
+		require.NoError(t, err, "failed to write file")
+		w.Close()
```

Comment:
https://pkg.go.dev/mime/multipart#Writer.Close it finishes the multipart message. If we use defer it would only do so when the test function returns.  This block is a test itself can we upload a DB. It's also a setup for further tests later on. So it needs to succeed before we can run the other subtests. That is also why it's not a subtest itself. As subtests can be run individually but they will need an uploaded DB 😅